tamr_toolbox/low_latency/llm.py

```diff
@@ -0,0 +1,215 @@
+import json
+import logging
+import time
+from collections import defaultdict
+from typing import Dict, List, Optional, Union
+
+from tamr_unify_client import Client
+from tamr_unify_client.operation import Operation
+
+from tamr_toolbox.models.data_type import JsonDict
+
+LOGGER = logging.getLogger(__name__)
+
+
+def update_llm_data(
+    client: Client,
+    *,
+    project_name: str,
+    do_update_clusters: bool = True,
+    do_use_manual_clustering: bool = False,
+) -> None:
+    """
+    Updates data for LLM query if needed, based on latest published clusters.
+
+    Args:
+        client: Tamr client object (connected to primary Tamr port , i.e. 9100)
+        project_name: name of the project to be updated
+        do_update_clusters: whether to update clusters, default True
+        do_use_manual_clustering: whether to use externally managed clustering, default False
+    """
+
+    url = (
+        f"projects/{project_name}:updateLLM?updateClusters={do_update_clusters}"
+        f"&useManualClustering={do_use_manual_clustering}"
+    )
+    response = client.post(url)
+    if not response.ok:
+        message = (
+            f"LLM update for {project_name} failed at submission time: "
+            + response.json()["message"]
+        )
+        LOGGER.error(message)
+        raise RuntimeError(message)
+    operation_id = response.content.decode("latin1")
+
+    # An operation id of '-1' is returned when LLM is already up to date
+    if operation_id != "-1":
+        operation = Operation.from_resource_id(client, operation_id)
+        operation.wait()
+    return None
+
+
+def poll_llm_status(
+    match_client: Client, *, project_name: str, num_tries: int = 10, wait_sec: int = 1
+) -> bool:
+    """
+    Check if LLM is queryable. Try up to num_tries times at 1s (or user-specified) interval.
```

Comment:
when would LLM not be queryable?

Reply:
If you haven't enabled LLM yet, it will definitely be false. My understanding is that for current versions of Tamr, it should never flip back to `False` after the LLM is enabled, but it sometimes does if something's gone wrong.

Also, for older versions of Tamr, LLM is not queryable while the LLM data is being updated (haven't tracked down which Tamr version made the change allow queries during updates, but it's a somewhat recent change).